File: Study Plan/LeetCode 75/0345 Reverse Vowels of a String/ReverseVowelsOfAString.java
/*  
    Problem
    --------------------------------------------------------------------------------
    345. Reverse Vowels of a String
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    - Runtime 8 ms (Beats 28.09% of users with Java)
    - Memory 44.90 MB (Beats 38.85% of users with Java)
*/

public class Solution {
    public String reverseVowels(String s) {
        StringBuilder text = new StringBuilder(s.length());
        Deque<Character> vowels = new ArrayDeque<>();
        for (int index = 0; index < s.length(); index++) {
            char letter = s.charAt(index);
            if (isVowel(letter)) {
                vowels.push(letter);
            }
        }
        for (int index = 0; index < s.length(); index++) {
            char letter = s.charAt(index);
            if(isVowel(letter)) {
                text.append(vowels.pop());
            } else {
                text.append(letter);
            }
        }
        return text.toString();
    }
    public boolean isVowel(char letter) {
        return "aeiouAEIOU".indexOf(letter) > -1;
    }
}

File: Study Plan/LeetCode 75/0345 Reverse Vowels of a String/ReverseVowelsOfAString.cs
/*  
    Problem
    --------------------------------------------------------------------------------
    345. Reverse Vowels of a String
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    https://leetcode.com/problems/reverse-vowels-of-a-string/solutions/4384067/c-linq-beats-91-in-runtime-time-and-space-complexity-o-n-stringbuilder-and-stack/
    
    - Runtime 58 ms (Beats 91.50% of users with C#)
    - Memory 44.92 MB (Beats 31.66% of users with C#)
*/

public class Solution
{
    public string ReverseVowels(string s)
    {
        var text = new StringBuilder(s.Length);
        var vowels = new Stack<char>();
        
        foreach (var letter in s)
        {
            if (isVowel(letter))
            {
                vowels.Push(letter);
            }
        }

        foreach (var letter in s)
        {
            if(isVowel(letter))
            {
                text.Append(vowels.Pop());
            }
            else
            {
                text.Append(letter);
            }
        }

        return text.ToString();
    }

    public bool isVowel(char letter)
    {
        return "aeiouAEIOU".Contains(letter);
    }
}
